File: app/src/main/java/ua/hospes/rtm/data/drivers/DriversRepositoryImpl.java
package ua.hospes.rtm.data.drivers;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Observable;
import io.reactivex.Single;
import ua.hospes.rtm.data.drivers.mapper.DriversMapper;
import ua.hospes.rtm.data.drivers.models.DriverDb;
import ua.hospes.rtm.data.drivers.storage.DriversDbStorage;
import ua.hospes.rtm.data.team.models.TeamDb;
import ua.hospes.rtm.data.team.storage.TeamsDbStorage;
import ua.hospes.rtm.domain.drivers.DriversRepository;
import ua.hospes.rtm.domain.drivers.models.Driver;
import ua.hospes.rtm.utils.Optional;

/**
 * @author Andrew Khloponin
 */
@Singleton
public class DriversRepositoryImpl implements DriversRepository {
    private final DriversDbStorage dbStorage;
    private final TeamsDbStorage   teamsDbStorage;


    @Inject
    DriversRepositoryImpl(DriversDbStorage dbStorage, TeamsDbStorage teamsDbStorage) {
        this.dbStorage = dbStorage;
        this.teamsDbStorage = teamsDbStorage;
    }


    @Override
    public Observable<Driver> get() {
        return dbStorage.get().flatMap(this::transform);
    }

    @Override
    public Observable<Driver> get(int... ids) {
        return dbStorage.get(ids).flatMap(this::transform);
    }

    @Override
    public Observable<Driver> getByTeamId(int teamId) {
        return dbStorage.getTeamById(teamId).flatMap(this::transform);
    }

    @Override
    public Observable<List<Driver>> listen() {
        return dbStorage.listen().flatMap(this::transform);
    }


    @Override
    public Observable<Boolean> save(Driver driver) {
        if (driver.getId() == -1) {
            return dbStorage.add(DriversMapper.map(driver)).map(result -> result.getResult() != 0);
        } else {
            return dbStorage.update(DriversMapper.map(driver)).map(result -> result.getResult() != 0);
        }
    }

    @Override
    public Single<Integer> remove(int id) {
        return dbStorage.remove(id);
    }

    @Override
    public Observable<Boolean> addDriversToTeam(int teamId, int... driverIds) {
        return dbStorage.addDriversToTeam(teamId, driverIds);
    }

    @Override
    public Observable<Boolean> removeDriversFromTeam(int teamId) {
        return dbStorage.removeDriversFromTeam(teamId);
    }

    @Override
    public Single<Integer> removeAll() {
        return dbStorage.removeAll();
    }


    private Observable<Driver> transform(DriverDb driverDb) {
        return Single.zip(Single.just(driverDb), getTeamById(driverDb.getTeamId()), DriversMapper::map).toObservable();
    }

    private Observable<List<Driver>> transform(List<DriverDb> driverDbs) {
        return Observable.fromIterable(driverDbs)
                .flatMapSingle(driverDb -> Single.zip(Single.just(driverDb), getTeamById(driverDb.getTeamId()), DriversMapper::map))
                .toList().toObservable();
    }

    private Single<Optional<TeamDb>> getTeamById(int id) {
        return teamsDbStorage.get(id).map(Optional::of).single(Optional.empty());
    }
}

File: app/src/main/java/ua/hospes/rtm/data/drivers/DriversRepositoryImpl.kt
package ua.hospes.rtm.data.drivers

import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.functions.BiFunction
import ua.hospes.rtm.data.team.TeamDb
import ua.hospes.rtm.data.team.TeamsDbStorage
import ua.hospes.rtm.domain.drivers.Driver
import ua.hospes.rtm.domain.drivers.DriversRepository
import ua.hospes.rtm.utils.Optional

class DriversRepositoryImpl(
        private val dbStorage: DriversDbStorage,
        private val teamsDbStorage: TeamsDbStorage
) : DriversRepository {

    override fun listen(): Observable<List<Driver>> = dbStorage.listen().flatMapSingle { this.transform(it) }

    override fun get(): Single<List<Driver>> = dbStorage.get().flatMapSingle { this.transform(it) }.toList()
    override fun get(vararg ids: Int): Single<List<Driver>> = dbStorage.get(*ids).flatMapSingle { this.transform(it) }.toList()
    override fun getByTeamId(teamId: Int): Single<List<Driver>> = dbStorage.getTeamById(teamId).flatMapSingle { this.transform(it) }.toList()


    override fun save(driver: Driver): Completable = when (driver.id) {
        null -> dbStorage.add(DriversMapper.map(driver)).map { it.result != 0L }.toList().ignoreElement()
        else -> dbStorage.update(DriversMapper.map(driver)).map { it.result != 0L }.toList().ignoreElement()
    }


    override fun addDriversToTeam(teamId: Int, vararg driverIds: Int): Completable = dbStorage.addDriversToTeam(teamId, *driverIds)

    override fun removeDriversFromTeam(teamId: Int): Completable = dbStorage.removeDriversFromTeam(teamId)
    override fun removeAll(): Completable = dbStorage.removeAll()
    override fun remove(id: Int): Completable = dbStorage.remove(id)


    private fun transform(driverDb: DriverDb) = Single.zip<DriverDb, Optional<TeamDb>, Driver>(
            Single.just(driverDb), getTeamById(driverDb.teamId!!),
            BiFunction<DriverDb, Optional<TeamDb>, Driver> { db, team -> DriversMapper.map(db, team) }
    )

    private fun transform(driverDbs: List<DriverDb>) = Observable.fromIterable(driverDbs)
            .flatMapSingle { transform(it) }.toList()

    private fun getTeamById(id: Int): Single<Optional<TeamDb>> =
            teamsDbStorage.get(id).map<Optional<TeamDb>> { Optional.of(it) }.single(Optional.empty())

}